lib/puppet/util/windows/monkey_patches/dir.rb

```diff
@@ -0,0 +1,416 @@
+      if attrib == INVALID_FILE_ATTRIBUTES ||
+         attrib & FILE_ATTRIBUTE_DIRECTORY == 0 ||
+         attrib & FILE_ATTRIBUTE_REPARSE_POINT == 0
+        bool = false
+      end
+
+      bool
+    end
+
+    alias :old_glob :glob
+
+    # Same as the standard MRI Dir.glob method except that it handles
+    # backslashes in path names.
+    #
+    def glob(glob_pattern, flags = 0, &block)
+      if glob_pattern.is_a?(Array)
+        temp = glob_pattern.map! { |pattern| string_check(pattern).tr("\\", "/") }
+      else
+        temp = string_check(glob_pattern).tr("\\", "/")
+      end
+
+      if flags.is_a?(Hash)
+        old_glob(temp, **flags, &block)
+      else
+        old_glob(temp, flags, &block)
+      end
+    end
+
+    alias old_ref []
+
+    # Same as the standard MRI Dir[] method except that it handles
+    # backslashes in path names.
+    #
+    def [](*glob_patterns)
+      temp = glob_patterns.map! { |pattern| "#{pattern}".tr("\\", "/") }
+      old_ref(*temp)
+    end
+
+    # JRuby normalizes the path by default.
+    unless RUBY_PLATFORM == "java"
+      alias oldgetwd getwd
+      alias oldpwd pwd
+
+      # Returns the present working directory. Unlike MRI, this method always
+      # normalizes the path.
+      #
+      # Examples:
+      #
+      #    Dir.chdir("C:/Progra~1")
+      #    Dir.getwd # => C:\Program Files
+      #
+      #    Dir.chdir("C:/PROGRAM FILES")
+      #    Dir.getwd # => C:\Program Files
+      #
+      def getwd
+        path1 = FFI::Buffer.new(:wint_t, 1024, true)
+        path2 = FFI::Buffer.new(:wint_t, 1024, true)
+        path3 = FFI::Buffer.new(:wint_t, 1024, true)
+
+        length = GetCurrentDirectoryW(path1.size, path1)
+
+        if length == 0 || length > path1.size
+          raise SystemCallError.new("GetCurrentDirectoryW", FFI.errno)
+        end
+
+        length = GetShortPathNameW(path1, path2, path2.size)
+
+        if length == 0 || length > path2.size
+          raise SystemCallError.new("GetShortPathNameW", FFI.errno)
+        end
+
+        length = GetLongPathNameW(path2, path3, path3.size)
+
+        if length == 0 || length > path3.size
+          raise SystemCallError.new("GetLongPathNameW", FFI.errno)
+        end
+
+        path = wstrip(path3.read_bytes(length * 2))
+
+        begin
+          path.encode(Encoding.default_external)
+        rescue Encoding::UndefinedConversionError
+          path.encode(Encoding::UTF_8)
+        end
+      end
+
+      alias :pwd :getwd
+      alias :reparse_dir? :junction?
+
+      # Simulate MRI's contortions for a stringiness check.
+      def string_check(arg)
+        return arg if arg.is_a?(String)
+        return arg.send(:to_str) if arg.respond_to?(:to_str, true) # MRI honors it, even if private
+        return arg.to_path if arg.respond_to?(:to_path)
+        raise TypeError
+      end
+
+      # Macro from Windows header file, used by the create_junction method.
+      def CTL_CODE(device, function, method, access)
```

Comment:
Should this be also private? same as win32-dir gem

Reply:
oops, missed those